include/boost/geometry/algorithms/detail/intersection/areal_areal.hpp

```diff
@@ -0,0 +1,348 @@
+// Boost.Geometry
+
+// Copyright (c) 2020, Oracle and/or its affiliates.
+// Contributed and/or modified by Adam Wulkiewicz, on behalf of Oracle
+
+// Licensed under the Boost Software License version 1.0.
+// http://www.boost.org/users/license.html
+
+#ifndef BOOST_GEOMETRY_ALGORITHMS_DETAIL_INTERSECTION_AREAL_AREAL_HPP
+#define BOOST_GEOMETRY_ALGORITHMS_DETAIL_INTERSECTION_AREAL_AREAL_HPP
+
+
+#include <boost/core/ignore_unused.hpp>
+
+#include <boost/geometry/algorithms/detail/intersection/interface.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace intersection
+{
+
+
+template
+<
+    typename GeometryOut,
+    typename OutTag = typename detail::intersection::tag
+                        <
+                            typename geometry::detail::output_geometry_value
+                                <
+                                    GeometryOut
+                                >::type
+                        >::type
+>
+struct intersection_areal_areal_
+{
+    template
+    <
+        typename Areal1,
+        typename Areal2,
+        typename RobustPolicy,
+        typename Strategy
+    >
+    static inline void apply(Areal1 const& areal1,
+                             Areal2 const& areal2,
+                             RobustPolicy const& robust_policy,
+                             GeometryOut& geometry_out,
+                             Strategy const& strategy)
+    {
+        geometry::dispatch::intersection_insert
+            <
+                Areal1, Areal2,
+                typename boost::range_value<GeometryOut>::type,
+                overlay_intersection
+            >::apply(areal1, areal2, robust_policy,
+                     geometry::range::back_inserter(geometry_out),
+                     strategy);
+    }
+};
+
+// TODO: Ideally this should be done in one call of intersection_insert
```

Comment:
This implementation for Areal/Areal calls `intersection_insert` for other combinations several times. This means that partition is called and then turns are analysed several times. However, overlay stuff called by the old/basic Areal/Areal implementation (when the output is MultiPolygon instead of tuple) has all of the information needed to generate the result in one call, so to generate Areal + Linear + PointLike output.

I decided to not modify it mainly because it could interfere with @barendgehrels work on rescaling. The code is also more complex than in case of Linear geometries, with enriching turns, storing states in turns and visitors which are not clear to me. So I'd prefer to leave it to @barendgehrels or do it when he's done with rescaling. Maybe rewrite it, maybe for C++11/14, etc. In short, it was simpler this way.